src/System.Console/src/System/ConsolePal.Windows.cs

```diff
@@ -600,51 +609,73 @@ public static int CursorTop
             }
         }
 
-        // Although msdn states that the max allowed limit is 65K,
-        // desktop limits this to 24500 as buffer sizes greater than it
-        // throw.
-        private const int MaxConsoleTitleLength = 24500;
-
         public static string Title
         {
             get
             {
-                string title = null;
-                int titleLength = -1;
-                int r = Interop.Kernel32.GetConsoleTitle(out title, out titleLength);
+                Span<char> initialBuffer = stackalloc char[256];
+                ValueStringBuilder builder = new ValueStringBuilder(initialBuffer);
 
-                if (0 != r)
+                do
                 {
-                    throw Win32Marshal.GetExceptionForWin32Error(r, string.Empty);
-                }
+                    uint result = Interop.Kernel32.GetConsoleTitleW(ref builder.GetPinnableReference(), (uint)builder.Capacity);
 
-                if (titleLength > MaxConsoleTitleLength)
-                    throw new InvalidOperationException(SR.ArgumentOutOfRange_ConsoleTitleTooLong);
+                    // The documentation asserts that the console's title is stored in a shared 64KB buffer.
+                    // The magic number that used to exist here (24500) is likely related to that.
+                    // A full UNICODE_STRING is 32K chars...
+                    Debug.Assert(result <= short.MaxValue, "shouldn't be possible to grow beyond UNICODE_STRING size");
 
-                Debug.Assert(title.Length == titleLength);
-                return title;
+                    if (result == 0)
+                    {
+                        int error = Marshal.GetLastWin32Error();
+                        switch (error)
+                        {
+                            case Interop.Errors.ERROR_INSUFFICIENT_BUFFER:
+                                // Typically this API truncates but there was a bug in RS2 so we'll make an attempt to handle
+                                builder.EnsureCapacity(builder.Capacity * 2);
+                                break;
+                            case Interop.Errors.ERROR_SUCCESS:
+                                return string.Empty;
+                            default:
+                                throw Win32Marshal.GetExceptionForWin32Error(error, string.Empty);
```

Comment:
In the ERROR_SUCCESS and default cases, should we be disposing builder?

Reply:
Success shouldn't be needed as we didn't grow out of the stack. The error case should be _super_ rare- I generally haven't been adding the cleanup code in these cases (partially based on feedback from Jan).

Reply:
> The error case should be super rare

If somebody gets a console with empty title, it can be hitting it 100%.

I think it is a good trade-off to ship the cleanup on paths that end up throwing exceptions; but I think the cleanup should be always done on non-exceptional paths, even if they are rare.

Reply:
You make this easier to reason about by having `return builder.ToString();` after the while loop.

Reply:
> If somebody gets a console with empty title, it can be hitting it 100%.
> I think the cleanup should be always done on non-exceptional paths, even if they are rare.

I agree. In the empty title case we haven't actually grabbed a buffer. If that was possible I'd absolutely free it.

> You make this easier to reason about by having return builder.ToString(); after the while loop.

It would be a little awkward given the switch statement here. Not sure how to break out of the while without adding complexity.

Reply:
```
           get
           {
                Span<char> initialBuffer = stackalloc char[256];
                ValueStringBuilder builder = new ValueStringBuilder(initialBuffer);

                while (true)
                {
                    uint result = Interop.Kernel32.GetConsoleTitleW(ref builder.GetPinnableReference(), (uint)builder.Capacity);

                    // The documentation asserts that the console's title is stored in a shared 64KB buffer.
                    // The magic number that used to exist here (24500) is likely related to that.
                    // A full UNICODE_STRING is 32K chars...
                    Debug.Assert(result <= short.MaxValue, "shouldn't be possible to grow beyond UNICODE_STRING size");

                    if (result == 0)
                    {
                        int error = Marshal.GetLastWin32Error();
                        switch (error)
                        {
                            case Interop.Errors.ERROR_SUCCESS:
                                break;
                            case Interop.Errors.ERROR_INSUFFICIENT_BUFFER:
                                // Typically this API truncates but there was a bug in RS2 so we'll make an attempt to handle
                                builder.EnsureCapacity(builder.Capacity * 2);
                                continue;
                            default:
                                throw Win32Marshal.GetExceptionForWin32Error(error, string.Empty);
                        }
                    }
                    else if (result >= builder.Capacity - 1 || (s_isWindows7 && result >= builder.Capacity / sizeof(char) - 1))
                    {
                        // Our buffer was full. As this API truncates we need to increase our size and reattempt.
                        // Note that Windows 7 copies count of bytes into the output buffer but returns count of chars
                        // and as such our buffer is only "half" its actual size.
                        //
                        // (If we're Windows 10 with a version lie to 7 this will be inefficient so we'll want to remove
                        //  this workaround when we no longer support Windows 7)
                        builder.EnsureCapacity(builder.Capacity * 2);
                        continue;
                    }

                    builder.Length = (int)result;
                    break;
                }

                return builder.ToString();
            }
```

Reply:
The same `while(true) ... continue;` flow is used in other places in this file.